golang/cmd/automigrate/migrations/migration.go

```diff
@@ -0,0 +1,258 @@
+	// If no, do nothing
+
+	if !checkForAnyMigration(db) {
+		applyMigrations(Semver{Major: 0, Minor: 0, Patch: 0}, current, db)
+	} else {
+		latest := getLatestMigration(db)
+		applyMigrations(latest, current, db)
+	}
+}
+
+func checkIfNewer(last Semver, current Semver) bool {
+	if last.Major < current.Major {
+		return false
+	}
+	if last.Major == current.Major && last.Minor < current.Minor {
+		return false
+	}
+	if last.Major == current.Major && last.Minor == current.Minor && last.Patch <= current.Patch {
+		return false
+	}
+	return true
+}
+func getLatestMigration(db *sql.DB) Semver {
+	result, err := db.Query("SELECT tomajor, tominor, topatch FROM migrationtable ORDER BY timestamp DESC LIMIT 1")
+	if err != nil {
+		zap.S().Fatalf("Error getting latest migration: %s", err)
+	}
+	var major, minor, patch int
+	err = result.Scan(&major, &minor, &patch)
+	if err != nil {
+		zap.S().Fatalf("Error getting latest migration: %s", err)
+	}
+	return Semver{Major: major, Minor: minor, Patch: patch}
+}
+
+func checkForAnyMigration(db *sql.DB) bool {
+	result, err := db.Query("SELECT count(1) FROM migrationtable")
+	if err != nil {
+		zap.S().Fatalf("Error checking for any migration: %s", err)
+	}
+	var count int
+	err = result.Scan(&count)
+	if err != nil {
+		zap.S().Fatalf("Error checking for any migration: %s", err)
+	}
+	return count > 0
+}
+
+func applyMigrations(last Semver, current Semver, db *sql.DB) {
+	if !checkIfNewer(last, current) {
+		zap.S().Infof("No migrations to apply")
+		return
+	}
+	zap.S().Infof("Applying migrations from %s to %s", last, current)
+
+	// Apply migrations
+	fullList := buildLinkedList()
+	applyList := SliceBetweenVersions(&fullList, last, current)
+	err := applyList.Apply(last, db)
+	if err != nil {
+		zap.S().Fatalf("Error applying migrations: %s", err)
+	}
+
+}
+
+type migrationfunc func(db *sql.DB) error
+
+type Node struct {
+	prev      *Node
+	next      *Node
+	migration migrationfunc
+	version   Semver
+}
+
+type List struct {
+	head *Node
+	tail *Node
+}
+
+func (L *List) Insert(mfunc migrationfunc, version Semver) {
+	list := &Node{
+		next:      L.head,
+		migration: mfunc,
+		version:   version,
+	}
+	if L.head != nil {
+		L.head.prev = list
+	}
+	L.head = list
+
+	l := L.head
+	for l.next != nil {
+		l = l.next
+	}
+	L.tail = l
+}
+
+func (L *List) Display() {
+	l := L.head
+	for l != nil {
+		fmt.Println(l.version)
+		l = l.next
+	}
+}
+
+func (L *List) Apply(oldVersion Semver, db *sql.DB) error {
+	if db == nil {
+		return errors.New("db cannot be nil")
+	}
+	if L.head == nil {
+		return errors.New("list is empty")
+	}
+	l := L.head
+	lastVersion := oldVersion
+	for l != nil {
+		zap.S().Infof("Applying migration %s", l.version)
+		err := l.migration(db)
+		if err != nil {
+			return err
+		}
+		// Set migration as applied
+		_, err = db.Exec(
+			"INSERT INTO migrationtable (frommajor, fromminor, frompatch, tomajor, tominor, topatch, timestamp) VALUES (?, ?, ?, ?, ?, ?, now())",
```

Comment:
Document in documentation :)

Reply:
See:
https://www.notion.so/united-manufacturing-hub/ab58fbca87834bfeb7f654af1dbb43fd?v=61b50b6953fd415bb2ac4dc5dccfc109&p=2d72668a371248ff801869f9fecf648f&pm=s

Reply:
Documented in upgrade guide